src/core/properties/positionproperty.cpp

```diff
@@ -28,115 +28,260 @@
-                                  : worldSpacePos);
+void PositionProperty::set(const vec3& pos, CoordinateSpace space, ApplyOffset applyOffset) {
+    NetworkLock lock(this);
+    referenceSpace_.set(space);
+    const vec3 offset{applyOffset == ApplyOffset::Yes ? getOffset() : vec3{0.0f}};
+    position_.set(pos + offset);
+}
+
+void PositionProperty::updatePosition(const vec3& pos, CoordinateSpace sourceSpace) {
+    position_.set(convertFromWorld(convertToWorld(pos, sourceSpace), referenceSpace_));
+}
+
+CoordinateSpace PositionProperty::getCoordinateSpace() const { return referenceSpace_; }
+
+vec3 PositionProperty::getWorldSpaceDirection() const {
+    if (camera_ &&
+        (referenceSpace_ == CoordinateSpace::View || referenceSpace_ == CoordinateSpace::Clip)) {
+        return convertToWorld(position_, referenceSpace_) - camera_->getLookTo();
+    }
+    return position_;
+}
+
+vec3 PositionProperty::convertFromWorld(const vec3& pos, CoordinateSpace targetSpace) const {
+    switch (targetSpace) {
+        case CoordinateSpace::World:
+            return pos;
+        case CoordinateSpace::View:
+            if (camera_) {
+                SpatialIdentity identity;
+                SpatialCameraCoordinateTransformerImpl transformer{identity, camera_->get()};
+                return vec3{transformer.getWorldToViewMatrix() * vec4{pos, 1.0f}};
+            } else {
+                return pos;
+            }
+        case CoordinateSpace::Clip:
+            if (camera_) {
+                SpatialIdentity identity;
+                SpatialCameraCoordinateTransformerImpl transformer{identity, camera_->get()};
+                vec4 clipCoords = transformer.getWorldToClipMatrix() * vec4{pos, 1.0f};
+                clipCoords /= clipCoords.w;
+                return vec3{clipCoords};
+            } else {
+                return pos;
+            }
+        default:
+            throw Exception(IVW_CONTEXT, "Unsupported coordinate space '{}'", targetSpace);
             break;
-        case Space::WORLD:
+    }
+    return vec3{0.0f};
+}
+
+vec3 PositionProperty::convertToWorld(const vec3& pos, CoordinateSpace sourceSpace) const {
+    switch (sourceSpace) {
+        case CoordinateSpace::World:
+            return pos;
+        case CoordinateSpace::View:
+            if (camera_) {
+                SpatialIdentity identity;
+                SpatialCameraCoordinateTransformerImpl transformer{identity, camera_->get()};
+                return vec3{transformer.getViewToWorldMatrix() * vec4{pos, 1.0f}};
+            } else {
+                return pos;
+            }
+        case CoordinateSpace::Clip:
+            if (camera_) {
+                SpatialIdentity identity;
+                SpatialCameraCoordinateTransformerImpl transformer{identity, camera_->get()};
+                return vec3{transformer.getClipToWorldMatrix() * vec4{pos, 1.0f}};
+            } else {
+                return pos;
+            }
```

Comment:
Would something like this not work? 
```suggestion
vec3 PositionProperty::convertToWorld(const vec3& pos, CoordinateSpace sourceSpace) const {
     if (!camera_) {
          return pos;
     }
     SpatialIdentity identity;
     SpatialCameraCoordinateTransformerImpl transformer{identity, camera_->get()};
     return vec3{transformer.getMatrix(sourceSpace, CoordinateSpace::World) * vec4{pos, 1.0f}};
}   
```

We might add a vec3 transformPosition(from, to, vec3 pos) to SpatialCoordinateTransformer 
and maybe a 
vec3 transformDirection(from, to, vec3 pos)

Reply:
good suggestion